gobblin-api/src/main/java/gobblin/publisher/DataPublisher.java

```diff
@@ -54,8 +82,7 @@ public abstract void publishMetadata(Collection<? extends WorkUnitState> tasks)
    * @param states is a {@link Collection} of {@link WorkUnitState}s.
    * @throws IOException if there is a problem with publishing the metadata or the data.
    */
-  public void publish(Collection<? extends WorkUnitState> states)
-      throws IOException {
+  public void publish(Collection<? extends WorkUnitState> states) throws IOException {
     publishMetadata(states);
```

Comment:
metadata doesn't include checkpoint. Checkpoints are published elsewhere (in `JobContext.persistDatasetState`.

I think metadata is published before data because in some implementation of data publisher, it needs to first publish the metadata (e.g., schema) before data can be published.